### PR TITLE
[SPARK-43611][SQL][PS][CONNCECT] Make `ExtractWindowExpressions` retain the `PLAN_ID_TAG`

### PR DESCRIPTION
**Pull request by zhengruifeng**
### What changes were proposed in this pull request?
Make rule `ExtractWindowExpressions` retain the `PLAN_ID_TAG `

### Why are the changes needed?
In https://github.com/apache/spark/pull/39925, we introduced a new mechanism to resolve expression with specified plan.

However, sometimes the plan ID might be discarded by some analyzer rules, and then some expressions can not be correctly resolved, this issue is the main blocker of PS on Connect.

### Does this PR introduce _any_ user-facing change?
yes, a lot of Pandas APIs enabled

### How was this patch tested?
Enable UTs

### CODE REVIEW

**Comment by zhengruifeng** [comment]
I have checked with @cloud-fan that we might have to modify the rules one by one.
The good news is that 90% UTs can be resolved by this single one, and I think we only need to touch another 3~4 rules.

@HyukjinKwon @itholic

**Comment by itholic** [comment]
> The good news is that 90% UTs can be resolved by this single one, and I think we only need to touch 3~4 more rules.

Great! Could you help creating tickets for remaining 3~4 more rules??

**Comment by zhengruifeng** [comment]
> > The good news is that 90% UTs can be resolved by this single one, and I think we only need to touch 3~4 more rules.
> 
> Great! Could you help creating tickets for remaining 3~4 more rules under [SPARK-42497](https://issues.apache.org/jira/browse/SPARK-42497)??

TBH, I am not very sure about which rules to modify, it needs further investigation

**Comment by itholic** [comment]
I got it. Just created single ticket here: SPARK-44492 for addressing undefined remaining tests so we don't miss it.

**Review by HyukjinKwon — APPROVED** [review]

**Comment by zhengruifeng** [comment]
I saw all tests in `pyspark-pandas-slow-connect` successed

image

not sure why it suddenly became
image

let me disable a few UT to see what happens

**Comment by zhengruifeng** [comment]
CI link: https://github.com/zhengruifeng/spark/actions/runs/5612996302/job/15208077409

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/trees/TreeNode.scala [hunk 1]
```diff
@@ -151,6 +151,12 @@ abstract class TreeNode[BaseType <: TreeNode[BaseType]]
     }
   }
 
+  def copyTagFrom[T](other: BaseType, tag: TreeNodeTag[T]): Unit = {
+    // copy a specified tag if exists
+    other.getTagValue(tag)
+      .foreach(this.setTagValue(tag, _))
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/trees/TreeNode.scala:157]:
```suggestion
    other.getTagValue(tag).foreach(this.setTagValue(tag, _))
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 2]
```diff
@@ -3124,7 +3124,9 @@ class Analyzer(override val catalogManager: CatalogManager) extends RuleExecutor
 
         // Finally, generate output columns according to the original projectList.
         val finalProjectList = aggregateExprs.map(_.toAttribute)
-        Project(finalProjectList, withWindow)
+        val newProject = Project(finalProjectList, withWindow)
+        newProject.copyTagFrom(f, LogicalPlan.PLAN_ID_TAG)
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:3128]: shall we copy all tags?

Comment (zhengruifeng) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:3128]: sounds good, let me have a try

**Comment by zhengruifeng** [comment]
cc @cloud-fan all tests passed, would you mind taking another look?

**Review by cloud-fan — APPROVED** [review]

**Comment by zhengruifeng** [comment]
thanks! merged to master

**Comment by itholic** [comment]
Thanks!

**Comment by HyukjinKwon** [comment]
Should we merge this to 3.5 too?

**Comment by zhengruifeng** [comment]
@HyukjinKwon not sure, maybe not needed?

**Pull request closed without merging**